r/R/dplyr.R

```diff
@@ -0,0 +1,92 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+select.RecordBatch <- function(.data, ...) {
+  .data$selected_columns <- c(.data$selected_columns, list(quos(...)))
+  .data
+}
+
+filter.RecordBatch <- function(.data, ..., .preserve = FALSE) {
+  .data$filtered_rows <- c(.data$filtered_rows, quos(...))
+  .data
+}
+
+collect.RecordBatch <- function(x, ...) {
+  filters <- evaluate_filters(x)
+  colnames <- names(x)
+  for (q in x$selected_columns) {
+    colnames <- vars_select(colnames, !!!q)
+  }
+  df <- as.data.frame(x[filters, colnames])
+  if (length(x$group_by_vars)) {
+    df <- dplyr::grouped_df(df, dplyr::groups(x)$group_names)
+  }
+  # Slight hack: since x is R6, each select/filter modified the object in place,
+  # which is not standard R behavior. Let's zero out x$selected_columns and
+  # x$filtered_rows and hope that this side effect cancels out the other
+  # unexpected side effects.
+  x$selected_columns <- list()
+  x$filtered_rows <- list()
+  dplyr::ungroup(x)
+  df
+}
+
+evaluate_filters <- function(x) {
+  if (length(x$filtered_rows) == 0) {
+    # Keep everything
+    return(TRUE)
+  }
+  # Cache the columns we need here
+  filter_data <- new.env()
```

Comment:
I would suggest rewriting this using rlang `env` functions — you already have an implicit dependency on rlang via vctrs, and using its API will avoid a bunch of subtle trickiness with the base api. (e.g. you probably should be specifying an explicit parent here)